ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+    let
+      db      = coerceDatabase db0
+      wrangle = either HD.UtxoValues HD.UtxoValues
+      go (LmdbBox k) (LmdbBox v) acc
+        | k <= last_excluded_key = acc
+        | otherwise = do
+            m <- acc
+            when (Map.size m >= count) $ Left m
+            pure $ Map.insert k v m
+    in
+      wrangle <$> LMDB.foldrWithKey go (Right Map.empty) db
+
+lmdbInitTable :: LedgerConstraint k v
+  => String
+  -> LMDB.Database k v
+  -> HD.UtxoValues k v
+  -> LMDB.Transaction LMDB.ReadWrite ()
+lmdbInitTable tbl_name db0 (HD.UtxoValues m) = do
+  let db = coerceDatabase db0
+  is_empty <- LMDB.null db
+  unless is_empty $ Exn.throw $ DbErrInitialisingNonEmpty tbl_name
+  void $ Map.traverseWithKey (\k v -> LMDB.put db (LmdbBox k) (Just (LmdbBox v))) m
+
+{-------------------------------------------------------------------------------
+  LMDB Interface specialized for ApplyMapKinds
+-------------------------------------------------------------------------------}
+
+getDb                  :: LMDB.Mode mode       =>                      NameMK   k v -> LMDB.Transaction mode           (LMDBMK   k v)
+initLMDBTable          :: LedgerConstraint k v =>        LMDBMK k v -> ValuesMK k v -> LMDB.Transaction LMDB.ReadWrite (EmptyMK  k v)
+writeLMDBTable         :: LedgerConstraint k v =>        LMDBMK k v -> DiffMK   k v -> LMDB.Transaction LMDB.ReadWrite (EmptyMK  k v)
+readLMDBTable          :: LedgerConstraint k v =>        LMDBMK k v -> KeysMK   k v -> LMDB.Transaction mode           (ValuesMK k v)
+initRangeReadLMDBTable :: LedgerConstraint k v => Int -> LMDBMK k v                 -> LMDB.Transaction mode           (ValuesMK k v)
+rangeReadLMDBTable     :: LedgerConstraint k v => Int -> LMDBMK k v -> KeysMK   k v -> LMDB.Transaction mode           (ValuesMK k v)
+
+getDb                                                  (NameMK   name) = LMDBMK name <$> LMDB.getDatabase (Just name)
+initLMDBTable                (LMDBMK tbl_name db) (ApplyValuesMK vals) = ApplyEmptyMK     <$  lmdbInitTable  tbl_name db vals
+writeLMDBTable               (LMDBMK _        db) (ApplyDiffMK   diff) = ApplyEmptyMK     <$  lmdbWriteTable          db diff
+readLMDBTable                (LMDBMK _        db) (ApplyKeysMK   keys) = ApplyValuesMK    <$> lmdbReadTable           db keys
+initRangeReadLMDBTable count (LMDBMK _        db)                      = ApplyValuesMK    <$> lmdbInitRangeReadTable count db
+rangeReadLMDBTable     count (LMDBMK _        db) (ApplyKeysMK   keys) = ApplyValuesMK    <$> lmdbRangeReadTable count db keys
+
+{-------------------------------------------------------------------------------
+ Db settings
+-------------------------------------------------------------------------------}
+
+readDbSettings ::
+     LMDB.Database () DbState
+  -> LMDB.Transaction mode DbState
+readDbSettings db = readDbSettingsMaybeNull db >>= \case
+  Just x -> pure x
+  -- FIXME: if this (and other errors) are thrown, the async thread will die but
```

Comment:
This is an important FIXME, probably worth fixing before merging. The 'happy' path however works fine.